frr-metrics/collector/bgp.go

```diff
@@ -0,0 +1,140 @@
+package collector
+
+import (
+	"encoding/json"
+	"os/exec"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const bgpConnected = "Established"
+
+var (
+	bgpNamespace = "metallb"
+	bgpSubsystem = "bgp"
+
+	sessionUpDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(bgpNamespace, bgpSubsystem, "session_up"),
+		"BGP session state (1 is up, 0 is down)",
+		[]string{
+			"peer",
+		},
+		nil,
+	)
+
+	updatesSentDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(bgpNamespace, bgpSubsystem, "updates_total"),
+		"Number of BGP UPDATE messages sent",
+		[]string{
+			"peer",
+		},
+		nil,
+	)
+
+	prefixesDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(bgpNamespace, bgpSubsystem, "announced_prefixes_total"),
+		"Number of prefixes currently being advertised on the BGP session",
+		[]string{
+			"peer",
+		},
+		nil,
+	)
+)
+
+type BGP struct {
+	Log          log.Logger
+	getNeighbors func() (string, error)
+}
+
+type FRRNeighbor struct {
+	BgpState string `json:"bgpState"`
+
+	MessageStats struct {
+		UpdatesSent int `json:"updatesSent"`
+	} `json:"messageStats"`
+
+	AddressFamilyInfo map[string]struct {
+		SentPrefixCounter int `json:"sentPrefixCounter"`
+	} `json:"addressFamilyInfo"`
+}
+
+func NewBGP() *BGP {
+	return &BGP{getNeighbors: getBGPNeighbors}
+}
+
+func (*BGP) Name() string {
+	return bgpSubsystem
+}
+
+func (c *BGP) Describe(ch chan<- *prometheus.Desc) {
+	describeNeighborMetrics(ch)
+}
+
+func describeNeighborMetrics(ch chan<- *prometheus.Desc) {
+	ch <- sessionUpDesc
+	ch <- updatesSentDesc
+	ch <- prefixesDesc
+}
+
+func (c *BGP) Collect(ch chan<- prometheus.Metric) {
+	res, err := c.getNeighbors()
+	if err != nil {
+		level.Error(c.Log).Log("error", err, "msg", "failed to fetch BGP neighbors from FRR")
+		return
+	}
+
+	neighbors, err := parseNeighbours(res)
```

Comment:
If I get this right, we are fetching the metrics in the very same moment prometheus asks us for them.
What I normally saw (but this doesn't necessarily mean that is the right way to do it), is that we put values in the GaugeVec (or other "bucket") and then prometheus scrapes the content of it.

Here, we are not event driven because we are pulling info from frr, but having a periodic goroutine that does the scraping and fills the buckets would make the trick. Prometheus scrape interval is normally long enough (like 30 seconds), so a goroutine that updates every 10 seconds would suffice I think.